doc/dev/cephx.rst

```diff
@@ -0,0 +1,353 @@
+- *principal_secret*: principal secret, a shared secret (16 bytes) known by the
+  principal and the authorizer
+- *mon_secret*: monitor secret, a shared secret known by all monitors
+- *service_secret*: a rotating secret known by all members of a
+  service class (e.g., all OSDs)
+
+- *auth ticket*: a ticket proving identity to the authorizers (monitors)
+- *service ticket*: a ticket proving identify and authorization to a service
+
+  
+Terminology
+-----------
+
+``{foo, bar}^secret`` denotes encryption by secret.
+
+
+Context
+-------
+
+The authentication messages described here are specific to the cephx
+auth implementation.  The messages are transferred by the Messenger
+protocol or by MAuth messages, depending on the version of the
+messenger protocol.  See also :ref:`msgr2`.
+
+An initial (messenger) handshake negotiates an authentication method to be used
+(cephx vs none or krb or whatever) and an assertion of what entity the client or
+daemon is attempting to authenticate as.
+
+Phase I: obtaining auth ticket
+------------------------------
+
+The cephx exchange begins with the monitor knowing who the client
+claims to be, and an initial cephx message from the authenticated/monitor to the
+client/principal.::
+
+  a->p : 
+    CephxServerChallenge {
+      u64 server_challenge     # random (by server)
+    }
+
+The client responds by adding its own challenge, and calculating a value derived
+from both challenges and its shared key principal_secret.::
+
+  p->a :
+    CephxRequestHeader {
+      u16 CEPHX_GET_AUTH_SESSION_KEY
+    }
+    CephXAuthenticate {
+      u8 1
+      u64 client_challenge     # random (by client)
+      u64 key = {client_challenge + server_challenge}^principal_secret   # (roughly)
+      blob old_ticket          # old ticket, if we are reconnecting or renewing
+    }
+   
+
+The authenticator/monitor looks up principal_secret in database, verifies key is correct.
+If old_ticket is present, verify it is valid, and we can reuse the same global_id.
+(Otherwise, a new global_id is assigned by the monitor.)::
+
+  a->p :
+    CephxReplyHeader {
+      u16 CEPHX_GET_AUTH_SESSION_KEY
+      s32 result (0)
+    }
+    u8 encoding_version = 1
+    u32 num_tickets ( = 1)
+    ticket_info           # (N = 1)
+
+where::
+
+    ticket_info {
+      u32 service_id       # CEPH_ENTITY_TYPE_AUTH
+      u8 msg_version (1)
+      {CephXServiceTicket service_ticket}^principal_secret
+      {CephxTicketBlob ticket_blob}^existing session_key   # if we are renewing a ticket,
+      CephxTicketBlob ticket_blob                          # otherwise
+    }
+
+    CephxServiceTicket {
+      CryptoKey session_key      # freshly generated (even if old_ticket is present)
+      utime_t expiration         # now + auth_mon_ticket_ttl
+    }
+
+    CephxTicketBlob {
+      u64 secret_id             # which service ticket encrypted this; -1 == monsecret
+      {CephXServiceTicketInfo ticket}^monsecret
```

Comment:
"monsecret" is not defined that I can see and I don't think we're using the shared mon secret key so I'm not sure what that's referring to? I see that we later refer to the client decrypting something with monsecret...
Does this just mean the client's secret key?

Reply:
mon_secret defined at top. added _

Reply:
There's a statement
>The principal validates the authorizer by decrypting the auth_ticket
>	with ``mon_secret`` and confirming that it says this principal is who they
>	say they are in the CephxAuthorizer fields.

Which doesn't make sense since the client (principal) doesn't see the mon_secret. I think maybe that should just say the monitor/authenticator validates the authorizer?

I got a bit confused switching between phase 2 and 3 trying to figure out what this was referencing so I didn't realize this struct was just for the monitors.